confusing error message around trait object bound

This example:

```rust
#![feature(in_band_lifetimes)]
#![feature(nll)]

trait DebugWith<Cx: ?Sized> {
    fn debug_with(&'me self, cx: &'me Cx) -> DebugCxPair<'me, Self, Cx> {
        DebugCxPair { value: self, cx }
    }
    
    fn fmt_with(&self, cx: &Cx, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result;
}

struct DebugCxPair<'me, Value: ?Sized, Cx: ?Sized>
where
    Value: DebugWith<Cx>,
{
    value: &'me Value,
    cx: &'me Cx,
}

trait DebugContext { }

struct Foo {
    bar: Bar
}

impl DebugWith<dyn DebugContext> for Foo {
    fn fmt_with(&self, cx: &dyn DebugContext, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let Foo { bar } = self;
        bar.debug_with(cx);
        Ok(())
    }
}

struct Bar { }

impl DebugWith<dyn DebugContext> for Bar {
    fn fmt_with(&self, cx: &dyn DebugContext, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Ok(())
    }
}

fn main() { }
```

gives this error:

```
error: unsatisfied lifetime constraints
  --> src/main.rs:29:9
   |
27 |     fn fmt_with(&self, cx: &dyn DebugContext, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
   |                 -          - let's call the lifetime of this reference `'1`
   |                 |
   |                 let's call the lifetime of this reference `'2`
28 |         let Foo { bar } = self;
29 |         bar.debug_with(cx);
   |         ^^^^^^^^^^^^^^^^^^ argument requires that `'1` must outlive `'2`
```

It took me some time to puzzle out what was happening here:

- `cx: &dyn DebugContext` expands to `&'a dyn (DebugContext + 'a)`
- the default from the *impl* however is `dyn (DebugContext + 'static)`
- the impl method is accepted because it is *more* general than the trait definition
- but it fails to recursively invoke

The error message without NLL, of course, is also not very illuminating.

I forgot to actual test how migration-mode (enabled for the 2018 edition) compares to `#![feature(nll)]`.

<details>

Current (nightly) 2018 edition output ([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=69a1ebb510ab314d4c16d4e095429f79)): 

```
error[E0495]: cannot infer an appropriate lifetime due to conflicting requirements
  --> src/main.rs:26:24
   |
26 |         bar.debug_with(cx);
   |                        ^^
   |
note: first, the lifetime cannot outlive the anonymous lifetime #2 defined on the method body at 24:5...
  --> src/main.rs:24:5
   |
24 | /     fn fmt_with(&self, cx: &dyn DebugContext, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
25 | |         let Foo { bar } = self;
26 | |         bar.debug_with(cx);
27 | |         Ok(())
28 | |     }
   | |_____^
note: ...so that the declared lifetime parameter bounds are satisfied
  --> src/main.rs:26:24
   |
26 |         bar.debug_with(cx);
   |                        ^^
   = note: but, the lifetime must be valid for the static lifetime...
   = note: ...so that the expression is assignable:
           expected &(dyn DebugContext + 'static)
              found &dyn DebugContext

error: aborting due to previous error
```

Current (nightly) output with `#![feature(nll)]` ([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7f9eea660e3fcaaf0a7cfbef3650b591)):

```
error: lifetime may not live long enough
  --> src/main.rs:28:24
   |
26 |     fn fmt_with(&self, cx: &dyn DebugContext, _fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
   |                            - let's call the lifetime of this reference `'1`
27 |         let Foo { bar } = self;
28 |         bar.debug_with(cx);
   |                        ^^ cast requires that `'1` must outlive `'static`
```

</details>

--- a/src/test/ui/nll/issue-54779-anon-static-lifetime.rs
+++ b/src/test/ui/nll/issue-54779-anon-static-lifetime.rs
@@ -1,0 +1,51 @@
+// Regression test for #54779, checks if the diagnostics are confusing.
+
+#![feature(nll)]
+
+trait DebugWith<Cx: ?Sized> {
+    fn debug_with<'me>(&'me self, cx: &'me Cx) -> DebugCxPair<'me, Self, Cx> {
+        DebugCxPair { value: self, cx }
+    }
+
+    fn fmt_with(&self, cx: &Cx, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result;
+}
+
+struct DebugCxPair<'me, Value: ?Sized, Cx: ?Sized>
+where
+    Value: DebugWith<Cx>,
+{
+    value: &'me Value,
+    cx: &'me Cx,
+}
+
+trait DebugContext {}
+
+struct Foo {
+    bar: Bar,
+}
+
+impl DebugWith<dyn DebugContext> for Foo {
+    fn fmt_with(
+        &self,
+        cx: &dyn DebugContext,
+        fmt: &mut std::fmt::Formatter<'_>,
+    ) -> std::fmt::Result {
+        let Foo { bar } = self;
+        bar.debug_with(cx); //~ ERROR: lifetime may not live long enough
+        Ok(())
+    }
+}
+
+struct Bar {}
+
+impl DebugWith<dyn DebugContext> for Bar {
+    fn fmt_with(
+        &self,
+        cx: &dyn DebugContext,
+        fmt: &mut std::fmt::Formatter<'_>,
+    ) -> std::fmt::Result {
+        Ok(())
+    }
+}
+
+fn main() {}

--- a/src/test/ui/nll/issue-54779-anon-static-lifetime.stderr
+++ b/src/test/ui/nll/issue-54779-anon-static-lifetime.stderr
@@ -1,0 +1,11 @@
+error: lifetime may not live long enough
+  --> $DIR/issue-54779-anon-static-lifetime.rs:34:24
+   |
+LL |         cx: &dyn DebugContext,
+   |             - let's call the lifetime of this reference `'1`
+...
+LL |         bar.debug_with(cx);
+   |                        ^^ cast requires that `'1` must outlive `'static`
+
+error: aborting due to previous error
+

